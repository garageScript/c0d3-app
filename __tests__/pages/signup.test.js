import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../graphql/queries/getApp'
import SIGNUP_USER from '../../graphql/queries/signupUser'
import SignupPage from '../../pages/signup'
import userEvent from '@testing-library/user-event'

describe('Signup Page', () => {
  const fakeEmail = 'fake@email.com'
  const fakeUsername = 'fakeusername'
  const fakeFirstName = 'fakefirstname'
  const fakeLastName = 'fakelastname'

  const fillOutSignupForm = async getByTestId => {
    const emailField = getByTestId('email')
    const usernameField = getByTestId('username')
    const firstNameField = getByTestId('firstName')
    const lastNameField = getByTestId('lastName')

    // the type event needs to be delayed so the Formik validations finish
    await userEvent.type(emailField, fakeEmail, { delay: 1 })
    await userEvent.type(usernameField, fakeUsername, { delay: 1 })
    await userEvent.type(firstNameField, fakeFirstName, { delay: 1 })
    await userEvent.type(lastNameField, fakeLastName, { delay: 1 })
  }

  test('Should render success component on success', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: SIGNUP_USER,
          variables: {
            firstName: fakeFirstName,
            lastName: fakeLastName,
            email: fakeEmail,
            username: fakeUsername
          }
        },
        result: {
          data: {
            signup: {
              success: true,
              username: fakeUsername,
              error: null
            }
          }
        }
      }
    ]

    const { container, getByTestId, getByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SignupPage />
      </MockedProvider>
    )

    const submitButton = getByTestId('submit')

    await fillOutSignupForm(getByTestId)
    fireEvent.click(submitButton)

    await waitFor(() => {
      expect(getByText('Account created successfully!')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })

  test('Should render errors on fail', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: SIGNUP_USER,
          variables: {
            firstName: fakeFirstName,
            lastName: fakeLastName,
            email: fakeEmail,
            username: fakeUsername
          }
        },
        result: {
          data: {
            signup: {
              success: false,
              username: null,
              error: null
            }
          }
        }
      }
    ]

    const { container, getByTestId, getByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SignupPage />
      </MockedProvider>
    )

    const submitButton = getByTestId('submit')

    await fillOutSignupForm(getByTestId)
    fireEvent.click(submitButton)

    await waitFor(() => {
      expect(
        getByText(
          'Server cannot be reached. Please try again. If this problem persists, please send an email to support@c0d3.com'
        )
      ).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })
})
