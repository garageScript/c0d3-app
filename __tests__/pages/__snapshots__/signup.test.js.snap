// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page Should render errors on fail 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Create Account
          </h1>
          <p
            class="card-text"
          />
          <div
            class="bg-light m-auto px-5 border-0"
            data-testid="error-message"
          >
            <h5
              class="text-danger"
            >
               Server cannot be reached. Please try again. If this problem persists, please send an email to support@c0d3.com
            </h5>
          </div>
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group "
            >
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="email"
                name="email"
                placeholder="Email address"
                type="email"
                value="fake@email.com"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="username"
                name="username"
                placeholder="Username"
                value="fakeusername"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="firstName"
                name="firstName"
                placeholder="First name"
                value="fakefirstname"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="lastName"
                name="lastName"
                placeholder="Last name"
                value="fakelastname"
              />
              <button
                class="btn btn-primary btn-lg btn-block mb-3"
                data-testid="submit"
                type="submit"
              >
                Create Account
              </button>
            </div>
          </form>
          <p
            class="text-black-50"
          >
            Already have an account?
             
            <a
              class="text-primary"
              href="/login"
            >
              Login
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;

exports[`Signup Page Should render success component on success 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <img
            alt="Green circle with a checkmark inside."
            class="mb-4"
            height="100px"
            src="/assets/curriculum/icons/checked.svg"
            width="100px"
          />
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Account created successfully!
          </h1>
          <p
            class="card-text"
          />
          <a
            class="btn btn-primary"
            href="/confirm/undefined"
          >
            Click here to set your password.
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;
